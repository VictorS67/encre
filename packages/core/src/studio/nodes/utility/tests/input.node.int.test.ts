import { setMaxListeners } from 'node:events';
import { describe, expect, jest, test } from '@jest/globals';

import { OpenAIChat } from '../../../../events/inference/chat/llms/openai/chat.js';
import { HumanMessage } from '../../../../events/input/load/msgs/human.js';
import { SystemMessage } from '../../../../events/input/load/msgs/system.js';
import { ChatPrompt } from '../../../../events/input/load/prompts/chat.js';
import { SubGraph } from '../../../graph.js';
import { BaseInput } from '../../../input.js';
import {
  GraphInputs,
  GraphProcessor,
  ProcessEvent,
} from '../../../processor.js';
import { globalNodeRegistry } from '../../../registration/nodes.js';
import { coerceToData } from '../../../utils/coerce.js';
import { PromptNode } from '../../input/prompt.node.js';

describe('test InputNodeImpl', () => {
  test('test using InputNode to listen user inputs', async () => {
    const chatPromptNode = globalNodeRegistry.createDynamicRaw(
      new ChatPrompt([
        new SystemMessage(
          'The following conversation is about writing fairy tales. You are given some topics, and your task is to write the fairy tale that based on those topics.'
        ),
      ])
    );

    const inputNode = globalNodeRegistry.createDynamicRaw(
      new BaseInput({ dataTypes: { prompt: ['string', 'chat-message[]'] } })
    );

    const openAIChatNode = globalNodeRegistry.createDynamicRaw(
      new OpenAIChat()
    );

    const graph = new SubGraph({
      nodes: [chatPromptNode, inputNode, openAIChatNode],
      connections: [
        {
          fromNodeId: chatPromptNode.id,
          fromPortName: 'prompt',
          toNodeId: inputNode.id,
          toPortName: 'prompt',
        },
        {
          fromNodeId: inputNode.id,
          fromPortName: 'prompt',
          toNodeId: openAIChatNode.id,
          toPortName: 'prompt',
        },
      ],
    });

    const processor = new GraphProcessor(graph);

    const graphInputs: GraphInputs = {};

    processor.on(
      'newAbortController',
      (fields: { controller: AbortController }) => {
        setMaxListeners(100, fields.controller.signal);
      }
    );

    processor.processGraph(graphInputs);

    const pEvents: ProcessEvent[] = [];
    const pEventTypes: {
      type: ProcessEvent['#type'];
      isSubProcessor: ProcessEvent['isSubProcessor'];
    }[] = [];

    const pEventToIgnore: Array<ProcessEvent['#type']> = [
      'trace',
      'error',
      'newAbortController',
      'graphError',
      'nodeError',
    ];

    for await (const event of processor.events()) {
      if (!pEventToIgnore.includes(event['#type'])) {
        pEvents.push(event);

        // Mock user providing input after the processor is requiring
        if (event['#type'] === 'requireInput') {
          processor.userInput(inputNode.id, {
            // This is the mocked user input to the InputNode
            prompt: coerceToData([
              ...(chatPromptNode as PromptNode).data.toChatMessages(),
              new HumanMessage('witch, princess'),
            ]),
          });
        }

        pEventTypes.push({
          type: event['#type'],
          isSubProcessor: event['isSubProcessor'],
        });
      }
    }

    expect(pEvents).toMatchSnapshot();
  });
});
