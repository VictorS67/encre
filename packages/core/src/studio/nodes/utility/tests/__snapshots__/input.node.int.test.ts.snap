// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test InputNodeImpl test using InputNode to listen user inputs 1`] = `
[
  {
    "#type": "processInQueue",
    "isSubProcessor": false,
    "node": {
      "data": {
        "_grp": 1,
        "_id": [
          "events",
          "input",
          "load",
          "prompts",
          "chat",
          "ChatPrompt",
        ],
        "_kwargs": {
          "messages": [
            {
              "_grp": 1,
              "_id": [
                "events",
                "input",
                "load",
                "msgs",
                "system",
                "SystemMessage",
              ],
              "_kwargs": {
                "content": "The following conversation is about writing fairy tales. You are given some topics, and your task is to write the fairy tale that based on those topics.",
              },
              "_type": "constructor",
            },
          ],
        },
        "_type": "constructor",
      },
      "id": "R6RJQPHJFGKDNV0ah",
      "inputs": {},
      "outputs": {
        "prompt": [
          "string",
          "chat-message[]",
        ],
      },
      "subType": "chat",
      "type": "prompt",
      "visualInfo": {
        "position": {
          "x": 0,
          "y": 0,
        },
        "size": {
          "height": 500,
          "width": 300,
        },
      },
    },
    "processorId": "JNYep09tEidvz_e0F",
  },
  {
    "#type": "nodeStart",
    "inputs": {},
    "isSubProcessor": false,
    "node": {
      "data": {
        "_grp": 1,
        "_id": [
          "events",
          "input",
          "load",
          "prompts",
          "chat",
          "ChatPrompt",
        ],
        "_kwargs": {
          "messages": [
            {
              "_grp": 1,
              "_id": [
                "events",
                "input",
                "load",
                "msgs",
                "system",
                "SystemMessage",
              ],
              "_kwargs": {
                "content": "The following conversation is about writing fairy tales. You are given some topics, and your task is to write the fairy tale that based on those topics.",
              },
              "_type": "constructor",
            },
          ],
        },
        "_type": "constructor",
      },
      "id": "R6RJQPHJFGKDNV0ah",
      "inputs": {},
      "outputs": {
        "prompt": [
          "string",
          "chat-message[]",
        ],
      },
      "subType": "chat",
      "type": "prompt",
      "visualInfo": {
        "position": {
          "x": 0,
          "y": 0,
        },
        "size": {
          "height": 500,
          "width": 300,
        },
      },
    },
    "processId": "Z7srYmtPgPPzLMuxk",
    "processorId": "JNYep09tEidvz_e0F",
  },
  {
    "#type": "nodeFinish",
    "isSubProcessor": false,
    "node": {
      "data": {
        "_grp": 1,
        "_id": [
          "events",
          "input",
          "load",
          "prompts",
          "chat",
          "ChatPrompt",
        ],
        "_kwargs": {
          "messages": [
            {
              "_grp": 1,
              "_id": [
                "events",
                "input",
                "load",
                "msgs",
                "system",
                "SystemMessage",
              ],
              "_kwargs": {
                "content": "The following conversation is about writing fairy tales. You are given some topics, and your task is to write the fairy tale that based on those topics.",
              },
              "_type": "constructor",
            },
          ],
        },
        "_type": "constructor",
      },
      "id": "R6RJQPHJFGKDNV0ah",
      "inputs": {},
      "outputs": {
        "prompt": [
          "string",
          "chat-message[]",
        ],
      },
      "subType": "chat",
      "type": "prompt",
      "visualInfo": {
        "position": {
          "x": 0,
          "y": 0,
        },
        "size": {
          "height": 500,
          "width": 300,
        },
      },
    },
    "outputs": {
      "prompt": {
        "type": "chat-message[]",
        "value": [
          {
            "_grp": 1,
            "_id": [
              "events",
              "input",
              "load",
              "msgs",
              "system",
              "SystemMessage",
            ],
            "_kwargs": {
              "content": "The following conversation is about writing fairy tales. You are given some topics, and your task is to write the fairy tale that based on those topics.",
            },
            "_type": "constructor",
          },
        ],
      },
    },
    "processId": "Z7srYmtPgPPzLMuxk",
    "processorId": "JNYep09tEidvz_e0F",
  },
  {
    "#type": "processInQueue",
    "isSubProcessor": false,
    "node": {
      "data": {
        "_grp": 1,
        "_id": [
          "studio",
          "input",
          "BaseInput",
        ],
        "_kwargs": {
          "data_types": {
            "prompt": [
              "string",
              "chat-message[]",
            ],
          },
        },
        "_type": "constructor",
      },
      "id": "DrDUEeEK@cSO079zx",
      "inputs": undefined,
      "outputs": {
        "prompt": [
          "string",
          "chat-message[]",
        ],
      },
      "subType": "input",
      "type": "input",
      "visualInfo": {
        "position": {
          "x": 0,
          "y": 0,
        },
        "size": {
          "height": 500,
          "width": 300,
        },
      },
    },
    "processorId": "JNYep09tEidvz_e0F",
  },
  {
    "#type": "nodeStart",
    "inputs": {
      "prompt": {
        "type": "chat-message[]",
        "value": [
          {
            "_grp": 1,
            "_id": [
              "events",
              "input",
              "load",
              "msgs",
              "system",
              "SystemMessage",
            ],
            "_kwargs": {
              "content": "The following conversation is about writing fairy tales. You are given some topics, and your task is to write the fairy tale that based on those topics.",
            },
            "_type": "constructor",
          },
        ],
      },
    },
    "isSubProcessor": false,
    "node": {
      "data": {
        "_grp": 1,
        "_id": [
          "studio",
          "input",
          "BaseInput",
        ],
        "_kwargs": {
          "data_types": {
            "prompt": [
              "string",
              "chat-message[]",
            ],
          },
        },
        "_type": "constructor",
      },
      "id": "DrDUEeEK@cSO079zx",
      "inputs": undefined,
      "outputs": {
        "prompt": [
          "string",
          "chat-message[]",
        ],
      },
      "subType": "input",
      "type": "input",
      "visualInfo": {
        "position": {
          "x": 0,
          "y": 0,
        },
        "size": {
          "height": 500,
          "width": 300,
        },
      },
    },
    "processId": "iFbCYBN_@zxrAycCT",
    "processorId": "JNYep09tEidvz_e0F",
  },
  {
    "#type": "requireInput",
    "callback": [Function],
    "isSubProcessor": false,
    "node": {
      "data": {
        "_grp": 1,
        "_id": [
          "studio",
          "input",
          "BaseInput",
        ],
        "_kwargs": {
          "data_types": {
            "prompt": [
              "string",
              "chat-message[]",
            ],
          },
        },
        "_type": "constructor",
      },
      "id": "DrDUEeEK@cSO079zx",
      "inputs": undefined,
      "outputs": {
        "prompt": [
          "string",
          "chat-message[]",
        ],
      },
      "subType": "input",
      "type": "input",
      "visualInfo": {
        "position": {
          "x": 0,
          "y": 0,
        },
        "size": {
          "height": 500,
          "width": 300,
        },
      },
    },
    "processId": "iFbCYBN_@zxrAycCT",
    "processorId": "JNYep09tEidvz_e0F",
    "promptContext": {
      "prompt": {
        "type": "chat-message[]",
        "value": [
          {
            "_grp": 1,
            "_id": [
              "events",
              "input",
              "load",
              "msgs",
              "system",
              "SystemMessage",
            ],
            "_kwargs": {
              "content": "The following conversation is about writing fairy tales. You are given some topics, and your task is to write the fairy tale that based on those topics.",
            },
            "_type": "constructor",
          },
        ],
      },
    },
  },
  {
    "#type": "nodeFinish",
    "isSubProcessor": false,
    "node": {
      "data": {
        "_grp": 1,
        "_id": [
          "studio",
          "input",
          "BaseInput",
        ],
        "_kwargs": {
          "data_types": {
            "prompt": [
              "string",
              "chat-message[]",
            ],
          },
        },
        "_type": "constructor",
      },
      "id": "DrDUEeEK@cSO079zx",
      "inputs": undefined,
      "outputs": {
        "prompt": [
          "string",
          "chat-message[]",
        ],
      },
      "subType": "input",
      "type": "input",
      "visualInfo": {
        "position": {
          "x": 0,
          "y": 0,
        },
        "size": {
          "height": 500,
          "width": 300,
        },
      },
    },
    "outputs": {
      "prompt": {
        "type": "chat-message[]",
        "value": [
          {
            "_grp": 1,
            "_id": [
              "events",
              "input",
              "load",
              "msgs",
              "system",
              "SystemMessage",
            ],
            "_kwargs": {
              "content": "The following conversation is about writing fairy tales. You are given some topics, and your task is to write the fairy tale that based on those topics.",
            },
            "_type": "constructor",
          },
          {
            "_grp": 1,
            "_id": [
              "events",
              "input",
              "load",
              "msgs",
              "human",
              "HumanMessage",
            ],
            "_kwargs": {
              "content": "witch, princess",
            },
            "_type": "constructor",
          },
        ],
      },
    },
    "processId": "iFbCYBN_@zxrAycCT",
    "processorId": "JNYep09tEidvz_e0F",
  },
  {
    "#type": "processInQueue",
    "isSubProcessor": false,
    "node": {
      "data": {
        "_grp": 1,
        "_id": [
          "events",
          "inference",
          "chat",
          "chatlms",
          "openai",
          "OpenAIChat",
        ],
        "_kwargs": {
          "additional_kwargs": {},
          "callbacks": undefined,
          "frequency_penalty": 0,
          "logit_bias": undefined,
          "max_tokens": 2048,
          "model": "gpt-3.5-turbo",
          "openai_api_key": {
            "_grp": 1,
            "_id": [
              "OPENAI_API_KEY",
            ],
            "_type": "secret",
          },
          "presence_penalty": 0,
          "response_format_type": undefined,
          "seed": undefined,
          "stop_words": undefined,
          "stream": false,
          "temperature": 1,
          "timeout": undefined,
          "top_p": 1,
          "user": undefined,
          "verbose": undefined,
        },
        "_type": "constructor",
      },
      "id": "hZjt4pB6hmJSKmXbk",
      "inputs": {
        "prompt": [
          "string",
          "chat-message[]",
        ],
      },
      "outputs": {
        "info": [
          "object",
          "unknown",
        ],
        "message": "chat-message",
        "output": "string",
        "tokenUsage": [
          "object",
          "unknown",
        ],
      },
      "subType": "openai",
      "type": "chatlm",
      "visualInfo": {
        "position": {
          "x": 0,
          "y": 0,
        },
        "size": {
          "height": 500,
          "width": 300,
        },
      },
    },
    "processorId": "JNYep09tEidvz_e0F",
  },
  {
    "#type": "processInQueue",
    "isSubProcessor": false,
    "node": {
      "data": {
        "_grp": 1,
        "_id": [
          "studio",
          "input",
          "BaseInput",
        ],
        "_kwargs": {
          "data_types": {
            "prompt": [
              "string",
              "chat-message[]",
            ],
          },
        },
        "_type": "constructor",
      },
      "id": "DrDUEeEK@cSO079zx",
      "inputs": undefined,
      "outputs": {
        "prompt": [
          "string",
          "chat-message[]",
        ],
      },
      "subType": "input",
      "type": "input",
      "visualInfo": {
        "position": {
          "x": 0,
          "y": 0,
        },
        "size": {
          "height": 500,
          "width": 300,
        },
      },
    },
    "processorId": "JNYep09tEidvz_e0F",
  },
  {
    "#type": "nodeStart",
    "inputs": {
      "prompt": {
        "type": "chat-message[]",
        "value": [
          {
            "_grp": 1,
            "_id": [
              "events",
              "input",
              "load",
              "msgs",
              "system",
              "SystemMessage",
            ],
            "_kwargs": {
              "content": "The following conversation is about writing fairy tales. You are given some topics, and your task is to write the fairy tale that based on those topics.",
            },
            "_type": "constructor",
          },
          {
            "_grp": 1,
            "_id": [
              "events",
              "input",
              "load",
              "msgs",
              "human",
              "HumanMessage",
            ],
            "_kwargs": {
              "content": "witch, princess",
            },
            "_type": "constructor",
          },
        ],
      },
    },
    "isSubProcessor": false,
    "node": {
      "data": {
        "_grp": 1,
        "_id": [
          "events",
          "inference",
          "chat",
          "chatlms",
          "openai",
          "OpenAIChat",
        ],
        "_kwargs": {
          "additional_kwargs": {},
          "callbacks": undefined,
          "frequency_penalty": 0,
          "logit_bias": undefined,
          "max_tokens": 2048,
          "model": "gpt-3.5-turbo",
          "openai_api_key": {
            "_grp": 1,
            "_id": [
              "OPENAI_API_KEY",
            ],
            "_type": "secret",
          },
          "presence_penalty": 0,
          "response_format_type": undefined,
          "seed": undefined,
          "stop_words": undefined,
          "stream": false,
          "temperature": 1,
          "timeout": undefined,
          "top_p": 1,
          "user": undefined,
          "verbose": undefined,
        },
        "_type": "constructor",
      },
      "id": "hZjt4pB6hmJSKmXbk",
      "inputs": {
        "prompt": [
          "string",
          "chat-message[]",
        ],
      },
      "outputs": {
        "info": [
          "object",
          "unknown",
        ],
        "message": "chat-message",
        "output": "string",
        "tokenUsage": [
          "object",
          "unknown",
        ],
      },
      "subType": "openai",
      "type": "chatlm",
      "visualInfo": {
        "position": {
          "x": 0,
          "y": 0,
        },
        "size": {
          "height": 500,
          "width": 300,
        },
      },
    },
    "processId": "1MsPOAePuEU6Xim3z",
    "processorId": "JNYep09tEidvz_e0F",
  },
  {
    "#type": "nodeFinish",
    "isSubProcessor": false,
    "node": {
      "data": {
        "_grp": 1,
        "_id": [
          "events",
          "inference",
          "chat",
          "chatlms",
          "openai",
          "OpenAIChat",
        ],
        "_kwargs": {
          "additional_kwargs": {},
          "callbacks": undefined,
          "frequency_penalty": 0,
          "logit_bias": undefined,
          "max_tokens": 2048,
          "model": "gpt-3.5-turbo",
          "openai_api_key": {
            "_grp": 1,
            "_id": [
              "OPENAI_API_KEY",
            ],
            "_type": "secret",
          },
          "presence_penalty": 0,
          "response_format_type": undefined,
          "seed": undefined,
          "stop_words": undefined,
          "stream": false,
          "temperature": 1,
          "timeout": undefined,
          "top_p": 1,
          "user": undefined,
          "verbose": undefined,
        },
        "_type": "constructor",
      },
      "id": "hZjt4pB6hmJSKmXbk",
      "inputs": {
        "prompt": [
          "string",
          "chat-message[]",
        ],
      },
      "outputs": {
        "info": [
          "object",
          "unknown",
        ],
        "message": "chat-message",
        "output": "string",
        "tokenUsage": [
          "object",
          "unknown",
        ],
      },
      "subType": "openai",
      "type": "chatlm",
      "visualInfo": {
        "position": {
          "x": 0,
          "y": 0,
        },
        "size": {
          "height": 500,
          "width": 300,
        },
      },
    },
    "outputs": {
      "info": {
        "type": "object",
        "value": {
          "finishReason": "stop",
        },
      },
      "message": {
        "type": "chat-message",
        "value": {
          "_grp": 1,
          "_id": [
            "events",
            "input",
            "load",
            "msgs",
            "assistant",
            "BotMessage",
          ],
          "_kwargs": {
            "additional_kwargs": {},
            "content": "Once upon a time, in a faraway kingdom, there lived a beautiful princess named Elara. She was known throughout the land for her kindness and grace. However, there was a shadow that loomed over the kingdom - a wicked witch named Malora who was determined to bring darkness and despair to the land.

Malora had her eye on the kingdom and coveted the throne for herself. She concocted a devious plan to capture the princess and take her place as the ruler of the kingdom. One fateful night, Malora cast a spell over the palace, causing a deep slumber to fall upon everyone inside, except for Elara.

As the princess wandered the empty halls of the palace, she came face to face with the witch. Malora cackled with glee, reveling in her victory as she revealed her plan to take over the kingdom. But Elara, filled with bravery and determination, refused to give in to the witch's dark desires.

With a flick of her wrist, Malora tried to cast a spell on Elara, but to her surprise, the princess was able to deflect it with a powerful force of light that radiated from within her. The witch realized that she had underestimated the princess and her inner strength.

As the two faced off in a battle of magic and willpower, Elara's goodness and light began to overpower the darkness that consumed Malora. With a final burst of magic, Elara broke the spell that had ensnared the kingdom and awakened her people from their slumber.

The kingdom rejoiced as they saw the evil witch defeated and their beloved princess standing strong and victorious. From that day on, Elara ruled the kingdom with wisdom and compassion, spreading light and love wherever she went.

And so, the tale of Princess Elara and the wicked witch Malora became a legend, reminding all who heard it of the power of goodness and the triumph of light over darkness.",
          },
          "_type": "constructor",
        },
      },
      "output": {
        "type": "string",
        "value": "Once upon a time, in a faraway kingdom, there lived a beautiful princess named Elara. She was known throughout the land for her kindness and grace. However, there was a shadow that loomed over the kingdom - a wicked witch named Malora who was determined to bring darkness and despair to the land.

Malora had her eye on the kingdom and coveted the throne for herself. She concocted a devious plan to capture the princess and take her place as the ruler of the kingdom. One fateful night, Malora cast a spell over the palace, causing a deep slumber to fall upon everyone inside, except for Elara.

As the princess wandered the empty halls of the palace, she came face to face with the witch. Malora cackled with glee, reveling in her victory as she revealed her plan to take over the kingdom. But Elara, filled with bravery and determination, refused to give in to the witch's dark desires.

With a flick of her wrist, Malora tried to cast a spell on Elara, but to her surprise, the princess was able to deflect it with a powerful force of light that radiated from within her. The witch realized that she had underestimated the princess and her inner strength.

As the two faced off in a battle of magic and willpower, Elara's goodness and light began to overpower the darkness that consumed Malora. With a final burst of magic, Elara broke the spell that had ensnared the kingdom and awakened her people from their slumber.

The kingdom rejoiced as they saw the evil witch defeated and their beloved princess standing strong and victorious. From that day on, Elara ruled the kingdom with wisdom and compassion, spreading light and love wherever she went.

And so, the tale of Princess Elara and the wicked witch Malora became a legend, reminding all who heard it of the power of goodness and the triumph of light over darkness.",
      },
      "tokenUsage": {
        "type": "object",
        "value": {
          "completionTokens": 392,
          "promptTokens": 44,
          "totalTokens": 436,
        },
      },
    },
    "processId": "1MsPOAePuEU6Xim3z",
    "processorId": "JNYep09tEidvz_e0F",
  },
  {
    "#type": "graphFinish",
    "graph": {
      "_grp": 1,
      "_id": [
        "studio",
        "graph",
        "SubGraph",
      ],
      "_kwargs": {
        "connections": [
          {
            "fromNodeId": "R6RJQPHJFGKDNV0ah",
            "fromPortName": "prompt",
            "toNodeId": "DrDUEeEK@cSO079zx",
            "toPortName": "prompt",
          },
          {
            "fromNodeId": "DrDUEeEK@cSO079zx",
            "fromPortName": "prompt",
            "toNodeId": "hZjt4pB6hmJSKmXbk",
            "toPortName": "prompt",
          },
        ],
        "nodes": [
          {
            "data": {
              "_grp": 1,
              "_id": [
                "events",
                "input",
                "load",
                "prompts",
                "chat",
                "ChatPrompt",
              ],
              "_kwargs": {
                "messages": [
                  {
                    "_grp": 1,
                    "_id": [
                      "events",
                      "input",
                      "load",
                      "msgs",
                      "system",
                      "SystemMessage",
                    ],
                    "_kwargs": {
                      "content": "The following conversation is about writing fairy tales. You are given some topics, and your task is to write the fairy tale that based on those topics.",
                    },
                    "_type": "constructor",
                  },
                ],
              },
              "_type": "constructor",
            },
            "id": "R6RJQPHJFGKDNV0ah",
            "inputs": {},
            "outputs": {
              "prompt": [
                "string",
                "chat-message[]",
              ],
            },
            "subType": "chat",
            "type": "prompt",
            "visualInfo": {
              "position": {
                "x": 0,
                "y": 0,
              },
              "size": {
                "height": 500,
                "width": 300,
              },
            },
          },
          {
            "data": {
              "_grp": 1,
              "_id": [
                "studio",
                "input",
                "BaseInput",
              ],
              "_kwargs": {
                "data_types": {
                  "prompt": [
                    "string",
                    "chat-message[]",
                  ],
                },
              },
              "_type": "constructor",
            },
            "id": "DrDUEeEK@cSO079zx",
            "inputs": undefined,
            "outputs": {
              "prompt": [
                "string",
                "chat-message[]",
              ],
            },
            "subType": "input",
            "type": "input",
            "visualInfo": {
              "position": {
                "x": 0,
                "y": 0,
              },
              "size": {
                "height": 500,
                "width": 300,
              },
            },
          },
          {
            "data": {
              "_grp": 1,
              "_id": [
                "events",
                "inference",
                "chat",
                "chatlms",
                "openai",
                "OpenAIChat",
              ],
              "_kwargs": {
                "additional_kwargs": {},
                "callbacks": undefined,
                "frequency_penalty": 0,
                "logit_bias": undefined,
                "max_tokens": 2048,
                "model": "gpt-3.5-turbo",
                "openai_api_key": {
                  "_grp": 1,
                  "_id": [
                    "OPENAI_API_KEY",
                  ],
                  "_type": "secret",
                },
                "presence_penalty": 0,
                "response_format_type": undefined,
                "seed": undefined,
                "stop_words": undefined,
                "stream": false,
                "temperature": 1,
                "timeout": undefined,
                "top_p": 1,
                "user": undefined,
                "verbose": undefined,
              },
              "_type": "constructor",
            },
            "id": "hZjt4pB6hmJSKmXbk",
            "inputs": {
              "prompt": [
                "string",
                "chat-message[]",
              ],
            },
            "outputs": {
              "info": [
                "object",
                "unknown",
              ],
              "message": "chat-message",
              "output": "string",
              "tokenUsage": [
                "object",
                "unknown",
              ],
            },
            "subType": "openai",
            "type": "chatlm",
            "visualInfo": {
              "position": {
                "x": 0,
                "y": 0,
              },
              "size": {
                "height": 500,
                "width": 300,
              },
            },
          },
        ],
      },
      "_type": "constructor",
    },
    "isSubProcessor": false,
    "outputs": {
      "hZjt4pB6hmJSKmXbk": {
        "info": {
          "type": "object",
          "value": {
            "finishReason": "stop",
          },
        },
        "message": {
          "type": "chat-message",
          "value": {
            "_grp": 1,
            "_id": [
              "events",
              "input",
              "load",
              "msgs",
              "assistant",
              "BotMessage",
            ],
            "_kwargs": {
              "additional_kwargs": {},
              "content": "Once upon a time, in a faraway kingdom, there lived a beautiful princess named Elara. She was known throughout the land for her kindness and grace. However, there was a shadow that loomed over the kingdom - a wicked witch named Malora who was determined to bring darkness and despair to the land.

Malora had her eye on the kingdom and coveted the throne for herself. She concocted a devious plan to capture the princess and take her place as the ruler of the kingdom. One fateful night, Malora cast a spell over the palace, causing a deep slumber to fall upon everyone inside, except for Elara.

As the princess wandered the empty halls of the palace, she came face to face with the witch. Malora cackled with glee, reveling in her victory as she revealed her plan to take over the kingdom. But Elara, filled with bravery and determination, refused to give in to the witch's dark desires.

With a flick of her wrist, Malora tried to cast a spell on Elara, but to her surprise, the princess was able to deflect it with a powerful force of light that radiated from within her. The witch realized that she had underestimated the princess and her inner strength.

As the two faced off in a battle of magic and willpower, Elara's goodness and light began to overpower the darkness that consumed Malora. With a final burst of magic, Elara broke the spell that had ensnared the kingdom and awakened her people from their slumber.

The kingdom rejoiced as they saw the evil witch defeated and their beloved princess standing strong and victorious. From that day on, Elara ruled the kingdom with wisdom and compassion, spreading light and love wherever she went.

And so, the tale of Princess Elara and the wicked witch Malora became a legend, reminding all who heard it of the power of goodness and the triumph of light over darkness.",
            },
            "_type": "constructor",
          },
        },
        "output": {
          "type": "string",
          "value": "Once upon a time, in a faraway kingdom, there lived a beautiful princess named Elara. She was known throughout the land for her kindness and grace. However, there was a shadow that loomed over the kingdom - a wicked witch named Malora who was determined to bring darkness and despair to the land.

Malora had her eye on the kingdom and coveted the throne for herself. She concocted a devious plan to capture the princess and take her place as the ruler of the kingdom. One fateful night, Malora cast a spell over the palace, causing a deep slumber to fall upon everyone inside, except for Elara.

As the princess wandered the empty halls of the palace, she came face to face with the witch. Malora cackled with glee, reveling in her victory as she revealed her plan to take over the kingdom. But Elara, filled with bravery and determination, refused to give in to the witch's dark desires.

With a flick of her wrist, Malora tried to cast a spell on Elara, but to her surprise, the princess was able to deflect it with a powerful force of light that radiated from within her. The witch realized that she had underestimated the princess and her inner strength.

As the two faced off in a battle of magic and willpower, Elara's goodness and light began to overpower the darkness that consumed Malora. With a final burst of magic, Elara broke the spell that had ensnared the kingdom and awakened her people from their slumber.

The kingdom rejoiced as they saw the evil witch defeated and their beloved princess standing strong and victorious. From that day on, Elara ruled the kingdom with wisdom and compassion, spreading light and love wherever she went.

And so, the tale of Princess Elara and the wicked witch Malora became a legend, reminding all who heard it of the power of goodness and the triumph of light over darkness.",
        },
        "tokenUsage": {
          "type": "object",
          "value": {
            "completionTokens": 392,
            "promptTokens": 44,
            "totalTokens": 436,
          },
        },
      },
    },
    "processorId": "JNYep09tEidvz_e0F",
  },
  {
    "#type": "done",
    "isSubProcessor": false,
    "processorId": "JNYep09tEidvz_e0F",
    "results": {
      "hZjt4pB6hmJSKmXbk": {
        "info": {
          "type": "object",
          "value": {
            "finishReason": "stop",
          },
        },
        "message": {
          "type": "chat-message",
          "value": {
            "_grp": 1,
            "_id": [
              "events",
              "input",
              "load",
              "msgs",
              "assistant",
              "BotMessage",
            ],
            "_kwargs": {
              "additional_kwargs": {},
              "content": "Once upon a time, in a faraway kingdom, there lived a beautiful princess named Elara. She was known throughout the land for her kindness and grace. However, there was a shadow that loomed over the kingdom - a wicked witch named Malora who was determined to bring darkness and despair to the land.

Malora had her eye on the kingdom and coveted the throne for herself. She concocted a devious plan to capture the princess and take her place as the ruler of the kingdom. One fateful night, Malora cast a spell over the palace, causing a deep slumber to fall upon everyone inside, except for Elara.

As the princess wandered the empty halls of the palace, she came face to face with the witch. Malora cackled with glee, reveling in her victory as she revealed her plan to take over the kingdom. But Elara, filled with bravery and determination, refused to give in to the witch's dark desires.

With a flick of her wrist, Malora tried to cast a spell on Elara, but to her surprise, the princess was able to deflect it with a powerful force of light that radiated from within her. The witch realized that she had underestimated the princess and her inner strength.

As the two faced off in a battle of magic and willpower, Elara's goodness and light began to overpower the darkness that consumed Malora. With a final burst of magic, Elara broke the spell that had ensnared the kingdom and awakened her people from their slumber.

The kingdom rejoiced as they saw the evil witch defeated and their beloved princess standing strong and victorious. From that day on, Elara ruled the kingdom with wisdom and compassion, spreading light and love wherever she went.

And so, the tale of Princess Elara and the wicked witch Malora became a legend, reminding all who heard it of the power of goodness and the triumph of light over darkness.",
            },
            "_type": "constructor",
          },
        },
        "output": {
          "type": "string",
          "value": "Once upon a time, in a faraway kingdom, there lived a beautiful princess named Elara. She was known throughout the land for her kindness and grace. However, there was a shadow that loomed over the kingdom - a wicked witch named Malora who was determined to bring darkness and despair to the land.

Malora had her eye on the kingdom and coveted the throne for herself. She concocted a devious plan to capture the princess and take her place as the ruler of the kingdom. One fateful night, Malora cast a spell over the palace, causing a deep slumber to fall upon everyone inside, except for Elara.

As the princess wandered the empty halls of the palace, she came face to face with the witch. Malora cackled with glee, reveling in her victory as she revealed her plan to take over the kingdom. But Elara, filled with bravery and determination, refused to give in to the witch's dark desires.

With a flick of her wrist, Malora tried to cast a spell on Elara, but to her surprise, the princess was able to deflect it with a powerful force of light that radiated from within her. The witch realized that she had underestimated the princess and her inner strength.

As the two faced off in a battle of magic and willpower, Elara's goodness and light began to overpower the darkness that consumed Malora. With a final burst of magic, Elara broke the spell that had ensnared the kingdom and awakened her people from their slumber.

The kingdom rejoiced as they saw the evil witch defeated and their beloved princess standing strong and victorious. From that day on, Elara ruled the kingdom with wisdom and compassion, spreading light and love wherever she went.

And so, the tale of Princess Elara and the wicked witch Malora became a legend, reminding all who heard it of the power of goodness and the triumph of light over darkness.",
        },
        "tokenUsage": {
          "type": "object",
          "value": {
            "completionTokens": 392,
            "promptTokens": 44,
            "totalTokens": 436,
          },
        },
      },
    },
  },
  {
    "#type": "finish",
    "isSubProcessor": false,
    "processorId": "JNYep09tEidvz_e0F",
  },
]
`;
